Check sysinfo for Cpuset cpu.shares and blkio

hqhq opened the pull request:
Carried: #14015

If kernel is compiled with CONFIG_FAIR_GROUP_SCHED disabled cpu.shares
doesn't exist.
If kernel is compiled with CONFIG_CFQ_GROUP_IOSCHED disabled blkio.weight
doesn't exist.
If kernel is compiled with CONFIG_CPUSETS disabled cpuset won't be
supported.

We need to handle these conditions by checking sysinfo and verifying them.

Signed-off-by: Zefan Li lizefan@huawei.com
Signed-off-by: Qiang Huang h.huangqiang@huawei.com

lizf-os commented:
LGTM

sorry I didn't follow my previous PR.

runcom commented:
LGTM

cpuguy83 commented on daemon/daemon_unix.go:
```diff
@@ -166,6 +166,11 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 		}
 	}
+	if hostConfig.CPUShares > 0 && !daemon.SystemConfig().CPUShares {
+		warnings = append(warnings, "Your kernel does not support CPU cfs shares. Shares discarded.")
```
Can we remove the `cfs` bit? It's pretty linux specific for something that can be shared with other unix-ish platforms.

cpuguy83 commented on daemon/daemon_unix.go:
```diff
@@ -166,6 +166,11 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 	}
+	if hostConfig.CPUShares > 0 && !daemon.SystemConfig().CPUShares {
+		warnings = append(warnings, "Your kernel does not support CPU cfs shares. Shares discarded.")
+		logrus.Warnf("Your kernel does not support CPU cfs shares. Shares discarded.")
```
Same here with the `cfs`

hqhq committed:
Check sysinfo for Cpuset cpu.shares and blkio

Carried: #14015

If kernel is compiled with CONFIG_FAIR_GROUP_SCHED disabled cpu.shares
doesn't exist.
If kernel is compiled with CONFIG_CFQ_GROUP_IOSCHED disabled blkio.weight
doesn't exist.
If kernel is compiled with CONFIG_CPUSETS disabled cpuset won't be
supported.

We need to handle these conditions by checking sysinfo and verifying them.

Signed-off-by: Zefan Li <lizefan@huawei.com>
Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@cpuguy83 Done.

calavera commented:
why are `checkCgroupBlkioInfo` and `checkCgroupCpusetInfo` returning pointers? can we return value structs since they are not meant to be changed?

calavera commented:
I'm going to merge this as it is. I'll follow up with a PR changing to values, so we can land this into the release.

calavera commented:
LGTM

hqhq commented:
@calavera Thanks. No special reasons for returning pointers, just followed other subsystems.
